'Attempting WinRM connection' - timeout for httpClient

I have problem with building VM with Packer, but I think the issue can be in winrm (that's why I create issue here, not in Packer repo)

My host: Windows 8.1
Packer version: 0.10.1

I want to build: https://github.com/StefanScherer/packer-windows/blob/my/windows_2016_docker.json (-only=virtualbox-iso)
Log in gist: https://gist.github.com/mwieczorek/83e6034fc41f06a8583bc3315646ab1b

Packer is "Attempting WinRM connection..." every 5 seconds. But after a few minutes it stops 'attempting'.
Provisioning of the machine continue, and after a while winrm is available (I'm able to Enter-PSSession from my host laptop to VM).

I checked network traffic between my laptop and VM (using Microsoft Message Analyzer - https://www.microsoft.com/en-us/download/details.aspx?id=44226). 
I see that last attempt was waiting for response about 76 seconds. So I started to wonder if it's related to some kind of timeout. (I attached session from that tool - [packer.zip](https://github.com/masterzen/winrm/files/260281/packer.zip) )

I cloned Packer from Github and looked at [http.go](https://github.com/mitchellh/packer/blob/master/vendor/github.com/masterzen/winrm/winrm/http.go) file.
I see there:
```
func Http_post(client *Client, request *soap.SoapMessage) (response string, err error) {
	httpClient := &http.Client{Transport: client.transport}
```

I don't know Go but I thought: "no timeout is set for httpClient, so I'll try to set it":

```
func Http_post(client *Client, request *soap.SoapMessage) (response string, err error) {
	httpClient := &http.Client{Transport: client.transport, Timeout:100*time.Second}
```
	

And it looks like it fixed the problem. See log in gist: https://gist.github.com/mwieczorek/b29a253cc38cdf1ed43e32b356b74816




So my question is: is the timeout not set on purpose?
Or maybe it should/can be set?

> The only reason I think we could end up waiting, is if there is a firewall running on the guest dropping winrm connections until it is properly configured.

I think it can be true. Packer is waiting when the winrm is not configured yet, but maybe some default firewall rules are. 

> if winrm wasn't active on the guest, then the connection wouldn't proceed and we would return immediately with an error (without waiting).

Yes, in logs (from gist) we have:
`2016/05/11 22:49:25 packer.exe: 2016/05/11 22:49:25 [ERROR] WinRM connection err: unknown error Post http://127.0.0.1:2841/wsman: EOF`
And then suddenly it's waiting...
So it can be firewall.

Anyway:
If you can add timeout I will be grateful.

--- a/winrm.go
+++ b/winrm.go
@@ -39,6 +39,7 @@ func main() {
 		cacert   string
 		gencert  bool
 		certsize string
+		timeout  string
 	)
 
 	flag.StringVar(&hostname, "hostname", "localhost", "winrm host")
@@ -50,6 +51,7 @@ func main() {
 	flag.StringVar(&cacert, "cacert", "", "CA certificate to use")
 	flag.BoolVar(&gencert, "gencert", false, "Generate x509 client certificate to use with secure connections")
 	flag.StringVar(&certsize, "certsize", "", "Priv RSA key between 512, 1024, 2048, 4096. Default :2048")
+	flag.StringVar(&timeout, "timeout", "0s", "connection timeout")
 
 	flag.Parse()
 
@@ -74,8 +76,9 @@ func main() {
 	} else {
 
 		var (
-			certBytes []byte
-			err       error
+			certBytes      []byte
+			err            error
+			connectTimeout time.Duration
 		)
 
 		if cacert != "" {
@@ -87,7 +90,10 @@ func main() {
 
 		cmd = flag.Arg(0)
 
-		endpoint := winrm.NewEndpoint(hostname, port, https, insecure, &certBytes)
+		connectTimeout, err = time.ParseDuration(timeout)
+		check(err)
+
+		endpoint := winrm.NewEndpointWithTimeout(hostname, port, https, insecure, &certBytes, connectTimeout)
 		client, err := winrm.NewClient(endpoint, user, pass)
 		check(err)
 

--- a/winrm/client.go
+++ b/winrm/client.go
@@ -56,6 +56,7 @@ func newTransport(endpoint *Endpoint) (*http.Transport, error) {
 		TLSClientConfig: &tls.Config{
 			InsecureSkipVerify: endpoint.Insecure,
 		},
+		ResponseHeaderTimeout: endpoint.Timeout,
 	}
 
 	if endpoint.CACert != nil && len(*endpoint.CACert) > 0 {

--- a/winrm/command_test.go
+++ b/winrm/command_test.go
@@ -2,8 +2,14 @@ package winrm
 
 import (
 	"bytes"
+	"fmt"
 	"io"
 	"io/ioutil"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"strconv"
 	"strings"
 	"sync"
 	"time"
@@ -171,4 +177,50 @@ func (s *WinRMSuite) TestCloseCommandStopsFetch(c *C) {
 	case <-time.After(1 * time.Second):
 		c.Log("no poll within one second, assuming fetch has stopped")
 	}
+}
+
+func (s *WinRMSuite) TestConnectionTimeout(c *C) {
+	count := 0
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		time.Sleep(2 * time.Second)
+		w.Header().Set("Content-Type", "application/soap+xml")
+		switch count {
+		case 0:
+			{
+				count = 1
+				fmt.Fprintln(w, executeCommandResponse)
+			}
+		case 1:
+			{
+				count = 2
+				fmt.Fprintln(w, outputResponse)
+			}
+		default:
+			{
+				fmt.Fprintln(w, doneCommandResponse)
+			}
+		}
+	}))
+	defer ts.Close()
+
+	// find port
+	url, err := url.Parse(ts.URL)
+	if err != nil {
+		c.Error(err)
+	}
+	host, port, err := net.SplitHostPort(url.Host)
+	if err != nil {
+		c.Error(err)
+	}
+	iport, err := strconv.Atoi(port)
+	if err != nil {
+		c.Error(err)
+	}
+
+	client, err := NewClient(&Endpoint{Host: host, Port: iport, Timeout: 1 * time.Second}, "Administrator", "v3r1S3cre7")
+	c.Assert(err, IsNil)
+
+	shell := &Shell{client: client, ID: "67A74734-DD32-4F10-89DE-49A060483810"}
+	_, err = shell.Execute("ipconfig /all")
+	c.Assert(err, ErrorMatches, ".*timeout.*")
 }

--- a/winrm/endpoint.go
+++ b/winrm/endpoint.go
@@ -1,6 +1,9 @@
 package winrm
 
-import "fmt"
+import (
+	"fmt"
+	"time"
+)
 
 // Endpoint struct holds configurations
 // for the server endpoint
@@ -10,6 +13,7 @@ type Endpoint struct {
 	HTTPS    bool
 	Insecure bool
 	CACert   *[]byte
+	Timeout  time.Duration
 }
 
 func (ep *Endpoint) url() string {
@@ -23,7 +27,7 @@ func (ep *Endpoint) url() string {
 	return fmt.Sprintf("%s://%s:%d/wsman", scheme, ep.Host, ep.Port)
 }
 
-// NewEndpoint returns new pointer to struct Endpoint
+// NewEndpoint returns new pointer to struct Endpoint, with a default 60s response header timeout
 func NewEndpoint(host string, port int, https bool, insecure bool, cert *[]byte) *Endpoint {
 	return &Endpoint{
 		Host:     host,
@@ -31,5 +35,18 @@ func NewEndpoint(host string, port int, https bool, insecure bool, cert *[]byte)
 		HTTPS:    https,
 		Insecure: insecure,
 		CACert:   cert,
+		Timeout:  60 * time.Second,
+	}
+}
+
+// NewEndpointWithTimeout returns a new Endpoint with a defined timeout
+func NewEndpointWithTimeout(host string, port int, https bool, insecure bool, cert *[]byte, timeout time.Duration) *Endpoint {
+	return &Endpoint{
+		Host:     host,
+		Port:     port,
+		HTTPS:    https,
+		Insecure: insecure,
+		CACert:   cert,
+		Timeout:  timeout,
 	}
 }

--- a/winrm/endpoint_test.go
+++ b/winrm/endpoint_test.go
@@ -1,6 +1,8 @@
 package winrm
 
 import (
+	"time"
+
 	. "gopkg.in/check.v1"
 )
 
@@ -12,4 +14,14 @@ func (s *WinRMSuite) TestEndpointUrlHttp(c *C) {
 func (s *WinRMSuite) TestEndpointUrlHttps(c *C) {
 	endpoint := &Endpoint{Host: "abc", Port: 123, HTTPS: true}
 	c.Assert(endpoint.url(), Equals, "https://abc:123/wsman")
+}
+
+func (s *WinRMSuite) TestEndpointWithDefaultTimeout(c *C) {
+	endpoint := NewEndpoint("test", 5585, false, false, nil)
+	c.Assert(endpoint.Timeout, Equals, 60*time.Second)
+}
+
+func (s *WinRMSuite) TestEndpointWithTimeout(c *C) {
+	endpoint := NewEndpointWithTimeout("test", 5585, false, false, nil, 120*time.Second)
+	c.Assert(endpoint.Timeout, Equals, 120*time.Second)
 }

